Support multiline Timelion queries

cjcenizal opened the pull request:
Supersedes https://github.com/elastic/kibana/pull/11164 with a cleaner history.

Addresses https://github.com/elastic/kibana/issues/8881 and https://github.com/elastic/kibana/issues/11449.

![image](https://cloud.githubusercontent.com/assets/1238659/26334596/fce0f166-3f1a-11e7-8cb4-f6b49a544086.png)

![image](https://cloud.githubusercontent.com/assets/1238659/26334599/ff892974-3f1a-11e7-9af2-08f9635fa755.png)

![image](https://cloud.githubusercontent.com/assets/1238659/26334602/01ebdcfc-3f1b-11e7-9d09-5da4b848d6af.png)

![image](https://cloud.githubusercontent.com/assets/1238659/26334604/03a29784-3f1b-11e7-8e45-6f82de0f0513.png)

![image](https://cloud.githubusercontent.com/assets/1238659/26334607/0554656c-3f1b-11e7-8322-b503a55f2565.png)

cjcenizal commented on src/core_plugins/timelion/public/chain.peg:
```diff
@@ -21,7 +25,7 @@
 
 start
- = tree:series {
+ = space? tree:series {
```
This allows white space before the initial es call:

```
         .es(*)
```

cjcenizal commented on src/core_plugins/timelion/public/chain.peg:
```diff
@@ -87,7 +91,7 @@ function_name
 
 function "function"
- = space? '.' name:function_name space? '(' space? arg_list:arg_list? space? ')' {
+ = space? '.' name:function_name space? '(' space? arg_list:arg_list? space? ')' space? {
```
This allows whitespace between chained functions:

```
.add(1)      .add(1)
           .add(1)
```

cjcenizal commented:
CC @snide

cjcenizal commented:
The first commit introduces UI Framework styles which support an expandable query bar. These styles aren't ultimately used by the changes to the Timelion UI in this PR. We can remove these UI Framework changes, but first I want to check with the Discovery team to make sure they won't actually be useful.

ppisljar commented on ui_framework/dist/ui_framework.css:
```diff
@@ -1915,6 +1915,40 @@ body {
+ * 3. Cap the height so that we can offer auto-completion suggestions below the input.
+ * 4. When the input is collapsed, we want text input to be truncated with an ellipsis.
+ */
+.kuiLocalSearchInput--expandable {
```
i see quite some duplication between this and `_local_search.scss` ... whats the idea behind this ?

cjcenizal replied:
This is the auto-generated CSS file, so we can ignore it during review.

ppisljar commented on src/core_plugins/timelion/public/directives/timelion_expression_input_helpers.js:
```diff
@@ -0,0 +1,105 @@
+  }
+}
+
+export function suggest(expression, caretOffset, functionList, Parser) {
```
could you fix this function so it also shows the suggestion after you start writing your arguments in ?
`.es` ... shows me the doc for es function with all the arguments
after i type `.es(` the arguments disappear which is supper annoying.

i understand if you want to do it in separate pr but it might be a small change ....
maybe storing last valid caret position and in just showing the doc for that ?

ppisljar commented on ui_framework/components/local_nav/_local_search.scss:
```diff
@@ -53,6 +58,38 @@
 }
 
 /**
+ * 1. When it expands, the search input should lie on top of the content.
```
i guess this comments are no longer completely relevant ?

ppisljar commented on ui_framework/components/local_nav/_local_search.scss:
```diff
@@ -53,6 +58,38 @@
+ * 2. Expand as the user enters text.
+ * 3. Cap the height so that we can offer auto-completion suggestions below the input.
+ * 4. When the input is collapsed, we want text input to be truncated with an ellipsis.
+ * 5. Prevent textarea from being resized.
```
not sure if you wanted to prevent from resizing textarea by height ... i can still expand it

ppisljar commented on ui_framework/components/local_nav/_local_search.scss:
```diff
@@ -53,6 +58,38 @@
 
 /**
+ * 1. When it expands, the search input should lie on top of the content.
+ * 2. Expand as the user enters text.
```
input does not auto grow for me

ppisljar commented on ui_framework/components/local_nav/_local_search.scss:
```diff
@@ -53,6 +58,38 @@
 /**
+ * 1. When it expands, the search input should lie on top of the content.
+ * 2. Expand as the user enters text.
+ * 3. Cap the height so that we can offer auto-completion suggestions below the input.
```
i can expand the textarea as far as i want (multiple pages) .. the cap does not seem to be working

ppisljar commented on ui_framework/components/local_nav/_local_search.scss:
```diff
@@ -53,6 +58,38 @@
+ * 1. When it expands, the search input should lie on top of the content.
+ * 2. Expand as the user enters text.
+ * 3. Cap the height so that we can offer auto-completion suggestions below the input.
+ * 4. When the input is collapsed, we want text input to be truncated with an ellipsis.
```
no ellipsis, just the scrollbars

ppisljar commented on ui_framework/dist/ui_framework.css:
```diff
@@ -1915,6 +1932,43 @@ body {
     /* 1 */ }
 
 /**
+ * 1. When it expands, the search input should lie on top of the content.
```
this seems to be duplicated code ? or ?

ppisljar commented on ui_framework/doc_site/src/views/local_nav/local_nav_example.js:
```diff
@@ -125,6 +128,29 @@ export default props => (
     </GuideSection>
```
irrelevant ... any idea why does github color this lines red ?

cjcenizal replied:
My guess is that it doesn't recognize JSX or, if it does, it's trying to highlight invalid syntax because of the closing tag at the top?

cjcenizal commented:
@ppisljar I updated the PR so the suggestions don't disappear until the expression's been completed. Can you take another look?

alexfrancoeur commented:
@cjcenizal I had a chance to take this PR for a spin. Some thoughts and comments below:

* I had attempted to copy and paste a lengthy Timelion request from the new [getting started guide](https://www.elastic.co/guide/en/kibana/current/timelion-math.html):
`.es(index=metricbeat*, timefield=@timestamp, metric=max:system.network.in.bytes).derivative().divide(1048576).lines(fill=2, width=1).color(green).label("Inbound traffic").title("Network traffic (MB/s)"), .es(index=metricbeat*, timefield=@timestamp, metric=max:system.network.out.bytes).derivative().multiply(-1).divide(1048576).lines(fill=2, width=1).color(blue).label("Outbound traffic").legend(columns=2, position=nw)`
When copy and pasting, I would have expected the Timelion query bar to dynamically expand for the content that was in it. 
<img width="1192" alt="screen shot 2017-05-24 at 8 17 57 am" src="https://cloud.githubusercontent.com/assets/25181823/26402850/ebd81648-4059-11e7-9a50-ea79a1f2d8ec.png">

* This is somewhat related and the result is shown above. If I save the workspace, return to it and click into a visualization - I only see the last portion of the expression. What would be nice is if I could see the first portion of the expression with a `...` or some other indicator that there is more.
`.es(index=metricbeat*, timefield=@timestamp, metric=max:system.network.in.bytes).derivative().divide(1048576).lines(fill=2, width=1)...`
If I then decide to click into this expression, the input box would automatically expand. What do you think?

* Multi-line editing can get a bit complicated. It won't be possible to create a query similar to the one in https://github.com/elastic/kibana/issues/8881 directly in this timelion editor. When I attempt to move to a new line, I actually submit the query to execute. I need to use shift+enter to move to a new line.
![may-24-2017 08-29-14](https://cloud.githubusercontent.com/assets/25181823/26403219/745e0ac6-405b-11e7-9b16-ac70ddaa2a45.gif)
We could switch the submit key combination to match dev console and be cmd+enter? That may be inconsistent with our other query bar though. The only other idea I had was to differentiate single line vs. multi-line editing. If in the multi-line edit "mode", do not treat enter as submit. 
* Is there any way we can get rid of that extra gray space between the query bar and the top menu?
* Whitespace fix seems to be working!
<img width="1035" alt="screen shot 2017-05-24 at 8 40 11 am" src="https://cloud.githubusercontent.com/assets/25181823/26403542/ac55669e-405c-11e7-9fa6-18bbed901d5a.png">

I'll continue to play around with this but please let me know you'd like to discuss any of these in more detail. Regardless of my feedback, what we have here is already a significant improvement for lengthy queries 😄

ppisljar commented:
@cjcenizal what about my other comments ?

cjcenizal commented:
After talking with @ppisljar, I just want to clarify that the original code he commented on has been removed from this branch and now lives in https://github.com/elastic/kibana/pull/11975.

ppisljar commented:
the auto expand/collapse functionality would make this much more user friendly imo. As far as i notice this is pretty much the biggest difference from the previous implementation with div. 

However looking in the future it would be nice if at some point this would be much more like the 'console' editor .... (seems the suggestions work well there as well) so i am having second thoughts on the textarea.

I agree with Alex that, in any case, we should make this feel more like the console (same key shortcuts)

cjcenizal commented:
This is great feedback, peeps! I'll implement the following:
* Change the input to be submitted on cmd+enter, to match Console. This seems like a good thing to teach the user, though we may need to surface some help content somewhere.
* Auto-expand the input when you click into it.
* When the input is collapsed and content is truncated, show some indicator in the UI.

cjcenizal commented:
OK, so I ran into a couple issues trying to get the "peeking" behavior @alexfrancoeur suggested. You can see my attempt in https://github.com/elastic/kibana/pull/11972/commits/95f602d95d513a3e6d3314b73d9efa941e0a17bb.

First of all, I think we want to disable this behavior once the user has resized the textarea, my assumption being that you resize the textarea so you can see its contents better and it'd be really annoying for it to expand and collapse as you focus and blur it. However, because there's no default "resize" event, I had to hack together a way of deriving whether the user has resized the textarea or not.

I got that figured out, but then there's a [ridiculous Chrome bug](https://bugs.chromium.org/p/chromium/issues/detail?id=94583) which prevents the textarea from being resized below the "peek" height, but _only if you resize it while its focused_. This only happens in Chrome and at first I thought I was going crazy. So not only does Chrome provide the incorrect UX, it does so consistently, because it will work fine if you resize the textarea _before_ you've clicked on it, but it will not work correctly if you resize it _after_ you've clicked on it.

So I figured if our goal is simply to teach users that they can enter multiline expressions, let's just use the UI to communicate that without getting too fancy:

![image](https://cloud.githubusercontent.com/assets/1238659/26430127/285dd878-40a0-11e7-971c-a9c973f6ae0c.png)

Thoughts?

@alexfrancoeur Let's wait on adjusting the empty gray area in the local nav until we've merged @snide's color contrast changes in https://github.com/elastic/kibana/pull/11610. It will be easier to make adjustments after we get those changes in because the local nav's height no longer corresponds to the height of the icon in the side nav. If we adjust the height of that local nav now, it will look a little weird and inconsistent, due to the way those two heights are related.

ppisljar commented:
@thomasneirynck i agree that doesn't need to be part of this PR, i was just saying that it would be nice that when we decide to do that we don't need to throw all this work away, so keeping that in mind here and not locking us down to much is important imo. 

@cjcenizal i like what you did there!

alexfrancoeur commented:
@cjcenizal regarding the console-like submit - this definitely makes it easier to add multiple lines, so ++ on this. Should we add "keyboard tips" in the docs here similar to console? I don't think we'd have as much but an indent option would be nice to have here as well

<img width="703" alt="screen shot 2017-05-25 at 11 42 29 am" src="https://cloud.githubusercontent.com/assets/25181823/26457763/52f31860-413f-11e7-8cdf-b1dc973bff4b.png">

Making the input box multiple lines makes it more obvious that you can enter a longer query, but I still prefer something along the lines of https://github.com/elastic/kibana/pull/11975 for the user experience. What about something like [this example](https://codepen.io/vsync/pen/frudD)?  We could add a limit of 5 - 10 rows for auto-expand but a user can always drag further if they'd like. If a user clicks on a timelion panel that has a long query or if a user copy and pastes a long query, the input box will update to the appropriate or default maximum height.  They would still have the ability to expand further by dragging and dropping. 

Also I think it's fine to wait for https://github.com/elastic/kibana/pull/11610 before improving the top menu.

cjcenizal commented:
@alexfrancoeur Thanks for the cool example! I would really like to implement this, but I think the Chrome bug I linked to will set us up for a poor and inconsistent UX. For example, open the link you shared in Chrome. Then enter multiple lines into the textarea. Then try to resize it smaller -- you can't. Imagine how frustrating it would be to paste or enter a multiline expression that pushes the graphs down, and being unable to narrow the textarea so you can see the graphs again. Now, refresh the page and resize the textarea. Enter multiple lines again. Now you can resize the textarea to be smaller, but only because you resized before you entered text.

I think we can hack around this bug with JavaScript, but I'd rather treat that as a separate enhancement in case I encounter problems. If you feel like this is a baseline satisfactory UX, then I suggest we merge it as-is and explore the improved UX in a separate PR.

cjcenizal commented:
@alexfrancoeur I added a keyboard tips section to the Help menu. Can you take a look?

ppisljar commented:
UX wise: this is possible to use and gives you something you didn't have before (multiline editor) ... but, the editor doesn't give you really good experience ... so not sure if this counts as 'baseline satisfactory UX' ... it does solve the issue, so it could be. I personally would love to see it improved, but i am having doubts if its worth the effort.

UI wise: this might look a bit 'less cool' that the rest of kibana (specially the new parts the design team is improving on !!) however it falls quite well into the whole timelion look&feel

alexfrancoeur commented:
@cjcenizal the keyboard section looks great! I'd almost say it's fair to use the when auto-complete is visible section as well. This seems to be applicable for timelion functions when auto-complete is triggered.

<img width="622" alt="screen shot 2017-05-26 at 8 57 45 am" src="https://cloud.githubusercontent.com/assets/25181823/26495431/8ca72602-41f1-11e7-85b6-3a5bc1ae20ed.png">

If the input box doesn't have the ability to shrink back then switching between panels can become a awkward. You could imagine switching between a 10 line expression and a single line expression. As I see it, this could behave in two ways. We either push the panels down based on the length of the query and transform the sheet each time you focus on a panel or we have a ton of empty space on the 1 line expression to keep the sheet consistent. Neither of which seem like a good solution to me.

I'm with Peter here. I think it stops the bleeding but isn't necessarily the best UX. This is my last pitch I swear 😄  What if we restricted this a bit more and made it an opt-in experience? Rather than allowing for a custom / dynamic input box, what if we defined a minimum and maximum height - giving the user the option to keep it collapsed or expand. To me, this gives me the feeling that I'm entering a workspace for creating timelion expressions.

The control is no good here but I thought this might help visualize my idea.
![collapsed](https://cloud.githubusercontent.com/assets/25181823/26496408/6e341690-41f5-11e7-8bf7-21282b67e299.png)

![expand](https://cloud.githubusercontent.com/assets/25181823/26496409/6e37f8a0-41f5-11e7-9dee-c654f02f1e59.png)

cjcenizal commented:
@thomasneirynck @ppisljar @snide I met up with @alexfrancoeur and we agreed that there's room for improvement in the UX, but because a) this PR provides functionality users have been asking for, b) the UX improvements depend upon a hack, and c) this PR targets and alpha release, we should merge this and get user feedback while I work on implementing the hack for improving the UX.

So my next steps will be: implementing Thomas's feedback and securing code review approvals from him and Peter, creating an issue to define the improved UX and identify the hack needed to implement it, and then merging this PR.

cjcenizal commented:
@thomasneirynck I've addressed your feedback, and @alexfrancoeur I added the Help tips you suggested:

![image](https://cloud.githubusercontent.com/assets/1238659/26506893/e67c2054-4202-11e7-85df-59f4ba0e823c.png)

Could you please re-review?

alexfrancoeur commented:
@cjcenizal LGTM - let's make sure to reference the issue that will tackle the improved UX here as well. Maybe update the description once it's created?

ddavidebor commented:
Thanks so much for this PR!
